lib/sqlalchemy/dialects/mysql/expression.py

```diff
@@ -0,0 +1,156 @@
+
+        SELECT id, firstname, lastname
+        FROM user
+        WHERE MATCH(firstname, lastname) AGAINST (:param_1 IN BOOLEAN MODE)
+        ORDER BY MATCH(firstname, lastname) AGAINST (:param_2) DESC
+
+    The :func:`.match_` function is a standalone version of the
+    :meth:`_expression.ColumnElement.match_` method available on all
+    SQL expressions, as when :meth:`_expression.ColumnElement.match_` is
+    used, but allows to pass multiple columns
+
+    All positional arguments passed to :func:`.match_`, typically should be a
+     :class:`_expression.ColumnElement` instances
+
+    :param: against typically scalar expression to be coerced into a ``str``
+
+    :param: flags optional ``dict``. Use properties ``in_boolean_mode``,
+     ``in_natural_language_mode`` and ``with_query_expansion`` to control it::
+
+        match_expr = match_(
+            users_table.c.firstname,
+            users_table.c.lastname,
+            against="John Connor",
+        )
+
+        print(match_expr)
+
+        # MATCH(firstname, lastname) AGAINST (:param_1)
+
+        print(match_expr.in_boolean_mode)
+
+        # MATCH(firstname, lastname) AGAINST (:param_1 IN BOOLEAN MODE)
+
+        print(match_expr.in_natural_language_mode.with_query_expansion)
+
+        # MATCH(firstname, lastname) AGAINST
+        # (:param_1 IN NATURAL LANGUAGE MODE WITH QUERY EXPANSION)
+
+    :property: ``in_boolean_mode`` returns new ``match_`` object with
+     set to ``True`` the ``mysql_boolean_mode`` flag
+
+    :property: ``in_natural_language_mode`` returns new ``match_`` object with
+     set to ``True`` the ``mysql_natural_language`` flag
+
+    :property: ``with_query_expansion`` returns new ``match_`` object with
+     set to ``True`` the ``mysql_query_expansion`` flag
+
+    .. versionadded:: 1.4.20
+
+    .. seealso::
+
+        :meth:`_expression.ColumnElement.match_`
+
+    """
+
+    __slots__ = (
+        "clause",
+        "against",
+        "flags",
+    )
+
+    default_flags = immutabledict(
+        {
+            "mysql_boolean_mode": False,
+            "mysql_natural_language": False,
+            "mysql_query_expansion": False,
+        }
+    )
+
+    def __init__(self, *clauselist, **kwargs):
+        clauselist_len = len(clauselist)
+        if clauselist_len == 0:
+            raise exc.CompileError("Can not match with no columns")
+        elif clauselist_len == 1:
+            self.clause = clauselist[0]
+        else:
+            clause = elements.BooleanClauseList._construct_raw(
+                operators.comma_op,
+                clauses=clauselist,
+            )
+            clause.group = False
+            self.clause = clause
+
+        against = kwargs.get("against")
+
+        if against is None:
+            raise exc.CompileError("Can not match without against")
+
+        self.against = against
+        self.flags = kwargs.get("flags", self.default_flags)
+
+    @property_enables_flag("mysql_boolean_mode")
```

Comment:
**mike bayer** (zzzeek) wrote:

using attribute access as mutators is surprising, and is inconsistent with how SQLAlchemy's other APIs work.  It's also more verbosity than we need for something simple like this.  While I may have illustrated how to do this for your own personal recipes, we can't have this style of mutation committed as part of the library.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2858

Reply:
It's not mutator. It's factory method, returning new `match_` instance. There are no surprises here. And test checking such surprises already exists.
What about inconsistency. I don't know how to make API as handy as possible. Explicit function calls? Do you see it in SQL? Explicit flags? Is it handy?

Reply:
hi  -

fine, not a "mutator", a "method chained" or "generator" method.  things that cause a change, like a new object to be explicitly created, are done with methods, not attributes.  an attribute access looks like a "get" and is not generally expected to be doing something new.

in any case, they are overkill here and are not consistent with how other SQLAlchemy look or work, so please remove them so we can merge this.

Reply:
Ok, let's go to tests of MatchExpression -- we can think the result API

Reply:
done